File: src/main/scala/clase6/ADTBundledStatefulSecure.scala
package clase6

object ADTBundledStatefulSecure extends App {

  case class Stack[T](push: T => Unit, pop: () => T, isEmpty: () => Boolean)

  def stack[T]():Stack[T] = {
    var content: List[T] = Nil

    def push(element:T) = {
      content = element :: content
    }

    def pop() = {
      content match {
        case head :: tail =>
          content = tail
          head
      }
    }

    def isEmpty() = content == Nil

    Stack(push, pop, isEmpty)
  }

  assert(stack().isEmpty())

  val s = stack[String]()

  s.push("ariel")
  s.push("alejandro")

  assert(!s.isEmpty())

  assert(s.pop() == "alejandro")
  assert(s.pop() == "ariel")
  assert(s.isEmpty())

}

File: haskell/src/clase6/ADTBundledStatefulSecure.hs
module ADTBundledStatefulSecure where

import Data.IORef

data Stack a = Stack {
  push :: a -> IO (),
  pop :: IO a,
  isEmpty :: IO Bool
}

push_ stack element = modifyIORef stack (element : )

pop_ stack = do
  current <- readIORef stack
  case current of
    (head : tail) -> do
      writeIORef stack tail
      return head

isEmpty_ stack = do
  current <- readIORef stack
  return (case current of
    [] -> True
    _ -> False)


stack = do
  content <- newIORef []
  return (Stack
    (push_ content)
    (pop_ content)
    (isEmpty_ content))

main = do
  s <- stack
  isEmpty s >>= print -- True
  push s "ariel"
  push s "alejandro"
  isEmpty s >>= print -- False
  pop s >>= print -- "alejandro"
  pop s >>= print -- "ariel"
  isEmpty s >>= print -- True


